x-pack/plugins/security/server/authentication/providers/base.ts

```diff
@@ -85,8 +91,13 @@ export abstract class BaseAuthenticationProvider {
    * @param [authHeaders] Optional `Headers` dictionary to send with the request.
    */
   protected async getUser(request: KibanaRequest, authHeaders: Headers = {}) {
-    return (await this.options.client
-      .asScoped({ headers: { ...request.headers, ...authHeaders } })
-      .callAsCurrentUser('shield.authenticate')) as AuthenticatedUser;
+    return deepFreeze({
+      ...(await this.options.client
+        .asScoped({ headers: { ...request.headers, ...authHeaders } })
+        .callAsCurrentUser('shield.authenticate')),
+      // We use `this.constructor` trick to get access to the static `type` field of the specific
```

Comment:
__note:__ couldn't find a better way :/

Reply:
I'm not aware of any other way to get access to a static property from an instance... The only other option that comes to mind is not making `getUser` return the `AuthenticatedUser` model, and instead return an interface which doesn't include the `authentication_provider` field. The `authenticator` itself could then add the provider which was used, and everything else could continue to use the AuthenticatedUser`.

Reply:
Yeah, that would be an option too, the tricky thing here is that `AuthenticationResult.user` that providers return is `AuthenticatedUser` and we rely on this fact in some other places... I tend to keep what I have for the time being and see how it goes.